features/steps/thens/alignment.py

```diff
@@ -172,3 +225,84 @@ def step_impl(context, inst):
                         yield ValidationOutcome(inst=inst, expected="same count of segments",
                                                 observed=observed_msg,
                                                 severity=OutcomeSeverity.ERROR)
+
+
+@gherkin_ifc.step(
+    'Each segment in the layout must have the same geometry type as its corresponding segment in the shape representation')
+def step_impl(context, inst):
+    # work back up the nesting tree to obtain the alignment
+    align = None
+    layout_entity_types = [
+        "IFCALIGNMENTHORIZONTAL",
+        "IFCALIGNMENTVERTICAL",
+        "IFCALIGNMENTCANT",
+    ]
+    for rel in inst.Nests:
+        layout = rel.RelatingObject
+        if layout.is_a().upper() in layout_entity_types:
+            for rel2 in layout.Nests:
+                align_ent = rel2.RelatingObject
+                if align_ent.is_a() == "IfcAlignment":
+                    align = ifc43x_alignment_validation.entities.Alignment().from_entity(align_ent)
+
+    if align is None:
+        msg = f"Error processing instance {str(inst)}. "
+        msg += "Expected an IfcAlignmentSegment nested 2 levels below an IfcAlignment."
+        raise ValueError(msg)
+
+    for idx, align_segment in enumerate(context.instances):
+
+        logic_segment = align_segment.DesignParameters
+        match logic_segment.is_a():
+            case "IfcAlignmentHorizontalSegment":
+                try:
+                    rep_segment = align.composite_curve.segments[idx].entity.ParentCurve
+                except AttributeError:
+                    try:
```

Comment:
Contrary to most pythonistas I'm not a big fan of try/except'ing for control flow. Could you please add a comment at least on which two different cases are handled here? ty

Reply:
Sure thing!  This is related to the previous comment where initially we try to locate the representations segment via index into the overall curve.  If that fails then we look at the representation attribute of the segment itself.  I'm leaning towards refactoring to ignore the overall curve and just look at the direct representation attribute.  Thoughts?

Reply:
Yes, that's probably a good way out, but I would break the rule into two distinct checks then:

- Correspondence of Segment --> ShapeRepresentation --> Item
- Segment geometry referencing an item that is also part of the overall composite curve.

Reply:
I actually ended up with 5 scenarios in the refactored `.feature` file:

1. Given IfcAlignment -> validate IfcCompositeCurve, IfcGradientCurve, or IfcSegmentedReferenceCurve as applicable
2. Given IfcAlignmentHorizontal -> validate IfcCompositeCurve
3. Given IfcAlignmentVertical -> validate IfcGradientCurve
4. Given IfcAlignmentCant -> validate IfcSegmentedReferenceCurve
5. Given IfcAlignmentSegment -> validate IfcCompositeCurveSegment or IfcCurveSegment as applicable

(Technically the steps in each scenario are the same after the initial given statement; this list is more pseudo code)